tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -312,6 +333,160 @@ private int WriteSpanLink(ref byte[] bytes, int offset, in SpanModel spanModel)
             return offset - originalOffset;
         }
 
+        private int WriteSpanEvent(ref byte[] bytes, int offset, in SpanModel spanModel)
+        {
+            int originalOffset = offset;
+
+            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _spanEventBytes);
+            offset += MessagePackBinary.WriteArrayHeader(ref bytes, offset, spanModel.Span.SpanEvents.Count);
+
+            foreach (var spanEvent in spanModel.Span.SpanEvents)
+            {
+                offset += MessagePackBinary.WriteMapHeader(ref bytes, offset, spanEvent.Attributes?.Count > 0 ? 3 : 2);
+
+                // time_unix_nano
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _timeUnixNanoBytes);
+                offset += MessagePackBinary.WriteInt64(ref bytes, offset, spanEvent.Timestamp.ToUnixTimeNanoseconds());
+
+                // name
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _nameBytes);
+                offset += MessagePackBinary.WriteString(ref bytes, offset, spanEvent.Name);
+
+                // attributes (strings only)
+                if (spanEvent.Attributes?.Count > 0)
+                {
+                    // Reserve space to patch the correct map header count later
+                    offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _attributesBytes);
+                    int attributeCountOffset = offset;
+                    offset += MessagePackBinary.WriteMapHeaderForceMap32Block(ref bytes, offset, 0); // placeholder
+
+                    int attrCount = 0;
+                    foreach (var attribute in spanEvent.Attributes)
+                    {
+                        if (string.IsNullOrEmpty(attribute.Key) || !IsAllowedAttributeType(attribute.Value))
```

Comment:
nit: There's a small difference between the allowed types between `IsAllowedAttributeType` and what you encode in `WriteEventAttribute` which should be fixed before merging. Here's the changes that need to be made:
- All types of unsigned integers are not accepted for span event attributes
- All types of signed integers should be written into the MessagePack as int64
- Both `float` and `double` should be written into the MessagePack as double, but I think we can safely ignore the `decimal` type

Reply:
Thank you so much for the details! I knew something needed to be changed and was waiting for a comment about it. done in: https://github.com/DataDog/dd-trace-dotnet/pull/6769/commits/324c4372c1d35f5b23bc9221643fab7d368516ae